EIPS/eip-165.md

```diff
@@ -0,0 +1,247 @@
+
+## Specification
+
+### How Interfaces are Identified
+
+For this standard, an *interface* is a set of [function selectors as defined by the Ethereum ABI](http://solidity.readthedocs.io/en/develop/abi-spec.html#function-selector). This a subset of [Solidity's concept of interfaces](http://solidity.readthedocs.io/en/develop/abi-spec.html) and the  `interface` keyword definition which also defines return types, mutability and events.
+
+We define the interface identifier as the XOR of all function selectors in the interface. This code example shows how to calculate an interface identifier:
+
+```solidity
+pragma solidity ^0.4.20;
+
+interface Solidity101 {
+    function hello() external pure;
+    function world(int) external pure;
+}
+
+contract Selector {
+    function calculateSelector() public pure returns (bytes4) {
+        Solidity101 i;
+        return i.hello.selector ^ i.world.selector;
+    }
+}
+```
+
+Note: interfaces do not permit optional functions, therefore, the interface identity will not include them.
+
+### How a Contract will Publish the Interfaces it Implements
+
+A contract that is compliant with ERC-165 shall implement the following interface (referred as `ERC165.sol`):
+
+```solidity
+pragma solidity ^0.4.20;
+
+interface ERC165 {
+    /// @notice Query if a contract implements an interface
+    /// @param interfaceID The interface identifier, as specified in ERC-165
+    /// @dev Interface identification is specified in ERC-165. This function
+    ///  uses less than 30,000 gas.
+    /// @return `true` if the contract implements `interfaceID` and
+    ///  `interfaceID` is not 0xffffffff, `false` otherwise
+    function supportsInterface(bytes4 interfaceID) external view returns (bool);
+}
+```
+
+The interface identifier for this interface is `0x01ffc9a7`. You can calculate this by running `        bytes4(keccak256('supportsInterface(bytes4)'));` or using the `Selector` contract above.
+
+Therefore the implementing contract will have a `supportsInterface` function that returns:
+
+- `true` when `interfaceID` is `0x01ffc9a7` (EIP165 interface)
+- `false` when `interfaceID` is `0xffffffff`
+- `true` for any other `interfaceID` this contract implements
+- `false` for any other `interfaceID`
+
+This function must return a bool and use at most 30,000 gas.
+
+Implementation note, there are several logical ways to implement this function. Please see the example implementations and the discussion on gas usage.
+
+### How to Detect if a Contract Implements ERC-165
+
+1. The source contact makes a `STATICCALL` to the destination address with input data: `0x01ffc9a701ffc9a7` and gas 30,000. This corresponds to `contract.supportsInterface("0x01ffc9a7")`.
```

Comment:
This (and two lines below) is likely what @chriseth refers to in [the comment](https://github.com/ethereum/EIPs/pull/881#issuecomment-366653818).

From [wiki](https://github.com/ethereum/wiki/wiki/Ethereum-Contract-ABI) (or [docs](https://solidity.readthedocs.io/en/develop/abi-spec.html?highlight=padded#formal-specification-of-the-encoding)) (emphasis mine):

> `bytes<M>`: `enc(X)` is the sequence of bytes in `X` **padded with zero-bytes to a length of 32**.

That page does not specify which end it should be padded from (different than the other occurences of the word "padded"; there also seems to be no example specifically for `bytes<M>` (or I missed it)). But other occurences that do the same pad on the left.

If so, the input data should be:

`0x01ffc9a70000000000000000000000000000000000000000000000000000000001ffc9a7`

(I might have gotten the zero count wrong.)

Reply:
Thank you, I think I get it. 734d68b6439941bf1577a4f1156b8e906185f7f5